src/lib.rs

```diff
@@ -198,6 +198,98 @@ impl_error!(bitcoin::hashes::hex::Error, Hex, Error);
 #[cfg(test)]
 mod test {
     use super::*;
+    use bitcoin::Amount;
+    use electrsd::{
+        bitcoind, bitcoind::bitcoincore_rpc::bitcoincore_rpc_json::AddressType,
+        bitcoind::bitcoincore_rpc::RpcApi, bitcoind::BitcoinD, ElectrsD,
+    };
+    use electrum_client::ElectrumApi;
+    use lazy_static::lazy_static;
+    use std::sync::{Mutex, Once};
+    use std::time::Duration;
+
+    lazy_static! {
+        static ref BITCOIND: BitcoinD = {
+            let bitcoind_exe =
+                bitcoind::downloaded_exe_path().expect("bitcoind version feature must be enabled");
+            let conf = bitcoind::Conf::default();
+            BitcoinD::with_conf(bitcoind_exe, &conf).unwrap()
+        };
+        static ref ELECTRSD: ElectrsD = {
+            let electrs_exe =
+                electrsd::downloaded_exe_path().expect("electrs version feature must be enabled");
+            let mut conf = electrsd::Conf::default();
+            conf.http_enabled = true;
+            ElectrsD::with_conf(electrs_exe, &BITCOIND, &conf).unwrap()
+        };
+        static ref MINER: Mutex<()> = Mutex::new(());
+    }
+
+    static PREMINE: Once = Once::new();
+
+    fn setup_clients() -> (BlockingClient, AsyncClient) {
+        PREMINE.call_once(|| {
+            let _miner = MINER.lock().unwrap();
+            generate_blocks_and_wait(101);
+        });
+
+        let esplora_url = ELECTRSD.esplora_url.as_ref().unwrap();
+
+        let builder = Builder::new(&format!("http://{}", esplora_url));
+        let blocking_client = builder.build_blocking().unwrap();
+
+        let builder_async = Builder::new(&format!("http://{}", esplora_url));
+        let async_client = builder_async.build_async().unwrap();
+
+        (blocking_client, async_client)
+    }
+
+    fn generate_blocks_and_wait(num: usize) {
+        let cur_height = BITCOIND.client.get_block_count().unwrap();
```

Comment:
I would move the line where you take the `MINER` lock in here, so that you don't have to remember to do it manually every time.

The main difference would be that the lock is dropped as soon as the blocks are generated, but I don't think any of your tests have explicit assertions on things like the block height, so I think it's fine if another tests generates blocks while the assertions are running

Reply:
No, in fact I had that variant before and moved the lock back into the individual methods after encountering some race conditions from the parallel tests (namely it couldn't be asserted that the results returned from the blocking/async clients would be equal, when other tests happen to create blocks in-between the calls).